jwst/extract_1d/soss_extract/pastasoss.py

```diff
@@ -0,0 +1,537 @@
+    for niter in range(maxiter):
+
+        # Assume all y have same wavelength.
+        wave_iterated = np.interp(x_grid - delta_x, x_dms[::-1],
+                                  wave_grid[::-1])  # Invert arrays to get increasing x.
+
+        # Compute the tilt angle at the wavelengths.
+        tilt_tmp = np.interp(wave_iterated, wave_grid, tilt)
+
+        # Compute the trace position at the wavelengths.
+        x_estimate = np.interp(wave_iterated, wave_grid, x_dms)
+        y_estimate = np.interp(wave_iterated, wave_grid, y_dms)
+
+        # Project that back to pixel coordinates.
+        x_iterated = x_estimate + (y_grid - y_estimate) * np.tan(np.deg2rad(tilt_tmp))
+
+        # Measure error between requested and iterated position.
+        delta_x = delta_x + (x_iterated - x_grid)
+
+        # If the desired precision has been reached end iterations.
+        if np.all(np.abs(x_iterated - x_grid) < dtol):
+            break
+
+    # Evaluate the final wavelength map, this time setting out-of-bounds values to NaN.
+    wave_map_2d = np.interp(x_grid - delta_x, x_dms[::-1], wave_grid[::-1], left=np.nan, right=np.nan)
+
+    # Extend to full detector size.
+    tmp = np.full((os * (dimx + 2 * xpad), os * (dimx + 2 * xpad)), fill_value=np.nan)
+    tmp[-os * (dimy + 2 * ypad):] = wave_map_2d
+    wave_map_2d = tmp
+
+    return wave_map_2d
+
+
+def get_soss_wavemaps(refmodel, pwcpos, subarray, padding=False, padsize=20, spectraces=False):
+    """
+    Generate order 1 and 2 2D wavemaps from the rotated SOSS trace positions
+
+    Parameters
+    ----------
+    pwcpos : float
+        The pupil wheel position
+    subarray: str
+        The subarray name, ['FULL', 'SUBSTRIP256', 'SUBSTRIP96']
+    padding : bool
+        Include padding on map edges (only needed for reference files)
+    padsize: int
+        The size of the padding to include on each side
+    spectraces : bool
+        Return the interpolated spectraces as well
+
+    Returns
+    -------
+    Array, Array
+        The 2D wavemaps and corresponding 1D spectraces
+    """
+    _, order1_x, order1_y, order1_wl = get_soss_traces(refmodel, pwcpos, order='1', subarray=subarray, interp=True)
+    _, order2_x, order2_y, order2_wl = get_soss_traces(refmodel, pwcpos, order='2', subarray=subarray, interp=True)
+
+    # Make wavemap from trace center wavelengths, padding to shape (296, 2088)
+    wavemin = 0.5
+    wavemax = 5.5
+    nwave = 5001
+    wave_grid = np.linspace(wavemin, wavemax, nwave)
+
+    # Extrapolate wavelengths for order 1 trace
+    xtrace_order1 = extrapolate_to_wavegrid(wave_grid, order1_wl, order1_x)
+    ytrace_order1 = extrapolate_to_wavegrid(wave_grid, order1_wl, order1_y)
+    spectrace_1 = np.array([xtrace_order1, ytrace_order1, wave_grid])
+
+    # Set cutoff for order 2 where it runs off the detector
+    o2_cutoff = 1783
+    w_o2_tmp = order2_wl[:o2_cutoff]
+    w_o2 = np.zeros(2040) * np.nan
```

Comment:
where does the 2040 come from?

Reply:
I believe this is the result of using the full frame, minus the 4 pixel reference pixel edges.